### PR TITLE
VariableCasingBear: Add VariableCasingBear

### PR DESCRIPTION
**Pull request by hypothesist**
Add VariableCaseBear and VariableCaseBearTest to the
bear library. Documentation under class definition.

### CODE REVIEW

**Comment by hypothesist** [comment]
Breaks because the coalang keywords PR in coala repo is not yet merged. Have to wait till then, I guess.

**Comment by AbdealiJK** [comment]
Is this blocked still ?
Can you put the issue/PR blocking it?

**Comment by sils1297** [comment]
@hypothesist can we continue work on this cool and awesome thing?

**Comment by hypothesist** [comment]
@sils1297 sorry for the late message, I thought we were waiting for @abhsag24 's work that would make this simpler. Or is that finished?

**Comment by hypothesist** [comment]
@AbdealiJK that issue was resolved. I was waiting for @abhsag24 to finish his bear helper that would simplify this. But the bear as such works alright even now.

**Comment by AbdealiJK** [comment]
Tests don't seem to pass - could you update appropriately ?

**Comment by hypothesist** [comment]
@AbdealiJK I'll restart work on this after a little while. That's alright?

**Comment by sils1297** [comment]
@hypothesist can we continue on this? I'd like to have this for the release.

**Comment by hypothesist** [comment]
@sils1297 Yep, this is right after caching is done.

File: bears/general/VariableCasingBear.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+            # We only want to change complete words, not abbreviations
+            # Example: abcDef becomes abc_def
+            # but abCDEf should be ab_CD_ef (better than ab_c_d_ef)
+            if line[i].isupper() and line[i - 1].islower():
+                rep += "_"
+                rep += line[i].lower()
+            else:
+                rep += line[i]
+            i += 1
+        return rep
```
Comment (Makman2) [inline bears/general/VariableCasingBear.py]: I'm using these functions too for my `FilenameBear` that formats filenames. I move this function into an own module, then you can rebase and use them from there too :)

File: bears/general/VariableCasingBear.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+                        vio,
+                        diffs={filename: diff},
+                        file=filename,
+                        line=first_line)
+                    result_texts = []
+        else:
+            self.warn("keywords and special_chars not defined for", language)
+            self.warn("Not running VariableCasingBear")
+
+    def all_to_camel(self, line):
```
Comment (Makman2) [inline bears/general/VariableCasingBear.py]: `line` is I think a bad name (it's not the line number and not the complete line, but the string you want to convert to camelcase) + you don't need self + what is `all`?

Comment (Makman2) [inline bears/general/VariableCasingBear.py]: got the algorithms here now:
https://github.com/coala-analyzer/coala/pull/2191/commits/aeff657a2fb44e98d0d2547f70dcc2152df5460b
:)
(though not merged yet)

Comment (Makman2) [inline bears/general/VariableCasingBear.py]: --> https://github.com/coala-analyzer/coala/pull/2191

Comment (hypothesist) [inline bears/general/VariableCasingBear.py]: I'll update in a few minutes.

File: bears/general/VariableCasingBear.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+
+class VariableCasingBear(LocalBear):
+
+    def run(self,
+            filename,
+            file,
+            casing: str,
+            language: str,
+            language_family: str,
```
Comment (abhsag24) [inline bears/general/VariableCasingBear.py]: `LanguageDefinition` doesn't need this anymore :P

**Comment by hypothesist** [comment]
Closing this pull request in favor of https://github.com/coala-analyzer/coala-bears/pull/484

(Since I somehow messed up my git stuff, also changed the whole thing, incorporating AnnotationsBear.)

**Pull request closed without merging**